src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,164 +1,581 @@
                 {
+                    if (!_map.TryGetValue(dagTemp, out var result))
+                    {
+                        // PROTOTYPE(patterns2): Not sure what temp kind should be used for `is pattern`.
+                        var temp = _factory.SynthesizedLocal(dagTemp.Type, syntax: _factory.Syntax, kind: SynthesizedLocalKind.SwitchCasePatternMatching);
+                        _map.Add(dagTemp, _factory.Local(temp));
+                        _temps.Add(temp);
+                        result = _factory.Local(temp);
+                    }
+
                     return result;
                 }
 
-                Debug.Assert((object)loweredPattern.Variable != null && loweredInput.Type.Equals(loweredPattern.Variable.GetTypeOrReturnType(), TypeCompareKind.AllIgnoreOptions));
+                public ImmutableArray<LocalSymbol> AllTemps()
+                {
+                    return _temps.ToImmutableArray();
+                }
 
-                var assignment = _factory.AssignmentExpression(loweredPattern.VariableAccess, loweredInput);
-                return _factory.MakeSequence(assignment, result);
+                public void AssignTemp(BoundDagTemp dagTemp, BoundExpression value)
+                {
+                    _map.Add(dagTemp, value);
+                }
             }
 
-            if (loweredPattern.VariableAccess.Kind == BoundKind.DiscardExpression)
+            private bool LowerDecision(BoundDagDecision decision)
             {
-                LocalSymbol temp;
-                BoundLocal discard = _factory.MakeTempForDiscard((BoundDiscardExpression)loweredPattern.VariableAccess, out temp);
-
-                return _factory.Sequence(ImmutableArray.Create(temp),
-                         sideEffects: ImmutableArray<BoundExpression>.Empty,
-                         result: MakeIsDeclarationPattern(loweredPattern.Syntax, loweredInput, discard, requiresNullTest: loweredInput.Type.CanContainNull()));
+                var oldSyntax = _factory.Syntax;
+                _factory.Syntax = decision.Syntax;
+                var result = LowerDecisionCore(decision);
+                _factory.Syntax = oldSyntax;
+                return result;
             }
 
-            return MakeIsDeclarationPattern(loweredPattern.Syntax, loweredInput, loweredPattern.VariableAccess, requiresNullTest: loweredInput.Type.CanContainNull());
-        }
-
-        /// <summary>
-        /// Is the test, produced as a result of a pattern-matching operation, always true?
-        /// Knowing that enables us to construct slightly more efficient code.
-        /// </summary>
-        private bool IsIrrefutablePatternTest(BoundExpression test)
-        {
-            while (true)
+            private bool LowerDecisionCore(BoundDagDecision decision)
             {
-                switch (test.Kind)
+                void addConjunct(ref IsPatternTranslator self, BoundExpression expression)
```

Comment:
Isn't `self` the same as `this`? Why pass it explicitly?

Reply:
An artifact of the refactoring that put the code here. I will eliminate it. Recorded in [https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222)

---
In reply to: [155097156](https://github.com/dotnet/roslyn/pull/23209#discussion_r155097156) [](ancestors = 155097156)